### PR TITLE
[IMP]  grid: add custom paste shortcuts

### PR DESCRIPTION
**Pull request by khpa-odoo**
This PR introduces following two shortcuts:
(i) Ctrl + D :

 -  Fills and overwrites selected cell(s) with the content and style of the cell above it in a column.
 -  Fills and overwrites multiple vertically selected cells in a column with the content and style of the anchor cell.

(i) Ctrl + R :

 -  Fills and overwrites selected cell(s) with the content and style of the cell on its left in a row.
 -  Fills and overwrites multiple horizontally selected cells in a row with the content and style of the anchor cell.

Task ID : 3476029

## Description:

description of this task, what is implemented and why it is implemented that way.

Task: : [TASK_ID](https://www.odoo.com/web#id=TASK_ID&action=333&active_id=2328&model=project.task&view_type=form&cids=1&menu_id=4720)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_lt("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

### CODE REVIEW

File: src/components/grid/grid.ts [hunk 1]
```diff
@@ -303,6 +303,52 @@ export class Grid extends Component<Props, SpreadsheetChildEnv> {
...
+        this.env.model.dispatch("COPY", { copyTarget });
+        await this.env.clipboard.write(this.env.model.getters.getClipboardContent());
+        this.env.model.dispatch("PASTE", { target: [zone] });
+      }
+    },
+    "CTRL+R": async () => {
+      if (this.env.model.getters.getSelectedZones().length === 1) {
+        const zone = this.env.model.getters.getSelectedZone();
+        const copyTarget =
+          zone.left !== zone.right
```
[the remark this answers is not in the document]
Comment (khpa-odoo) [reply src/components/grid/grid.ts:329]: Right, my bad. 
Applied the changes :+1:

File: src/components/grid/grid.ts [hunk 2]
```diff
@@ -303,6 +303,44 @@ export class Grid extends Component<Props, SpreadsheetChildEnv> {
...
+        this.env.model.dispatch("COPY", { copyTarget: [target] });
+        await this.env.clipboard.write(this.env.model.getters.getClipboardContent());
+        interactivePaste(this.env, [zone]);
+      }
+    },
+    "CTRL+R": async () => {
+      if (this.env.model.getters.getSelectedZones().length === 1) {
+        const zone = this.env.model.getters.getSelectedZone();
+        const target =
+          zone.left !== zone.right
```
Comment (LucasLefevre) [inline src/components/grid/grid.ts:329]: is `isSingleCellOrMerge`  what you want?

File: src/components/grid/grid.ts [hunk 3]
```diff
@@ -303,6 +303,44 @@ export class Grid extends Component<Props, SpreadsheetChildEnv> {
...
+                ...zone,
+                bottom: zone.top,
+              }
+            : {
+                ...zone,
+                top: zone.top - 1,
+                bottom: zone.top - 1,
+              };
+        this.env.model.dispatch("COPY", { copyTarget: [target] });
+        await this.env.clipboard.write(this.env.model.getters.getClipboardContent());
```
[the remark this answers is not in the document]
Comment (khpa-odoo) [reply src/components/grid/grid.ts]: Noted!
Applied your suggested approach. Thanks :+1:

File: src/helpers/ui/paste_interactive.ts [hunk 4]
```diff
@@ -8,7 +8,7 @@ export const PasteInteractiveContent = {
     "This operation is not possible due to a merge. Please remove the merges first than try again."
   ),
   wrongFigurePasteOption: _t("Cannot do a special paste of a figure."),
-  frozenPaneOverlap: _t("Cannot paste merged cells over a frozen pane."),
+  frozenPaneOverlap: _t("This operation is not allowed due to an overlapping frozen pane."),
```
[the remark this answers is not in the document]
Comment (khpa-odoo) [reply src/helpers/ui/paste_interactive.ts]: I made it generalised because while pasting non-merge cells by overlapping a frozen pane also displayed the same msg and I thought it would be misleading for user to display msg with reference to 'merged cells' when it is not the case.

Anyways, I've reverted the change for now :+1:

[the remark this answers is not in the document]
Comment (rrahir) [reply src/helpers/ui/paste_interactive.ts]: Actually you are 100% correct, I just stopped my search when I saw the code in the merge plugin. I completely misread the use case. the change you made completely makes sense. Sorry :/

[the remark this answers is not in the document]
Comment (khpa-odoo) [reply src/helpers/ui/paste_interactive.ts]: Okay I'll change it again and push once you're done with reviewing the rest part as well :+1:

[the remark this answers is not in the document]
Comment (khpa-odoo) [reply src/helpers/ui/paste_interactive.ts:11]: Changed it back to generalised form like the rest :)

File: src/types/commands.ts [hunk 5]
```diff
@@ -502,6 +502,7 @@ export interface UpdateLocaleCommand {
 // ------------------------------------------------
 export interface CopyCommand {
   type: "COPY";
+  copyTarget?: Zone[];
```
[the remark this answers is not in the document]
Comment (hokolomopo) [reply src/types/commands.ts]: I do'nt agree. I changed the "target" argument from cut to cutTarget, because "target" is not a keyword in commands and should always be a array of zones (not optional). We can argue whether or not we should allow optional, but it just makes the typing/generics functions easier.

And in the long/medium term we should make this argument disappear in CUT, the only reason it's needed is because we use CUT/PASTE to move columns, but we should really not and either have separate command or use the ClipboardState directly like Alexis.

File: src/components/grid/grid.ts [hunk 6]
```diff
@@ -303,6 +303,40 @@ export class Grid extends Component<Props, SpreadsheetChildEnv> {
...
+          zone.top !== zone.bottom
+            ? {
+                ...zone,
+                bottom: zone.top,
+              }
+            : {
+                ...zone,
+                top: zone.top - 1,
+                bottom: zone.top - 1,
+              };
```
Comment (hokolomopo) [inline src/components/grid/grid.ts:319]:
```suggestion
        const singleRowSelected  = zone.top !== zone.bottom
        const target ={
                ...zone,
                bottom: singleRowSelected ? zone.top : zone.top - 1,
                top: singleRowSelected ? zone.top : zone.top - 1,
              }
```
shorter and way easier to read IMO

same for the other shortcut

**Review by hokolomopo — COMMENTED** [review]
I'm not too sure about the commands. Their target is kinda useless, I guess the clipboard isn't the worst place to handle them but I'm not sure it should be its responsability, and they should definetly not mess up the clipboard state.

File: src/types/commands.ts [hunk 7]
```diff
@@ -516,6 +516,16 @@ export interface PasteCommand {
   pasteOption?: ClipboardPasteOptions;
 }
 
+export interface ReplicateCellsAboveCommand {
+  type: "REPLICATE_CELLS_ABOVE";
+  specificTarget?: Zone[];
```
[the remark this answers is not in the document]
Comment (khpa-odoo) [reply src/types/commands.ts]: Okay I've removed target from command and computed it in the plugin itself :+1:

File: src/plugins/ui_stateful/clipboard.ts [hunk 8]
```diff
@@ -109,6 +110,17 @@ export class ClipboardPlugin extends UIPlugin {
           this.status = "invisible";
         }
         break;
+      case "REPLICATE_CELLS_ABOVE":
+      case "REPLICATE_CELLS_ON_LEFT":
+        const zone = this.getters.getSelectedZone();
+        const copyTarget = ("specificTarget" in cmd && cmd.specificTarget) || [zone];
+        this.state = this.getClipboardState(copyTarget, "COPY");
```
[the remark this answers is not in the document]
Comment (khpa-odoo) [reply src/plugins/ui_stateful/clipboard.ts]: Okay so at present, I've pushed your suggestions which were on common grounds with the approach suggested 
by `rrahir`.
But I figure there's a contradiction or dilemma regarding introduction of specific commands as well as for using clipboard state directly like `laa`.

So I'm currently awaiting a final conclusion on desired approach from here on. Let me know which one of your approaches would be preferable and I'd then write my fix accordingly.
@rrahir @hokolomopo

File: src/plugins/ui_stateful/clipboard.ts [hunk 9]
```diff
@@ -109,6 +109,40 @@ export class ClipboardPlugin extends UIPlugin {
...
+      case "REPLICATE_CELLS_ABOVE":
+        if (this.getters.getSelectedZones().length === 1) {
+          const zone = this.getters.getSelectedZone();
+          const singleRowSelected = zone.top !== zone.bottom;
+          const copyTarget = {
+            ...zone,
+            bottom: singleRowSelected ? zone.top : zone.top - 1,
+            top: singleRowSelected ? zone.top : zone.top - 1,
+          };
+          this.state = this.getClipboardState([copyTarget], "COPY");
```
Comment (hokolomopo) [inline src/plugins/ui_stateful/clipboard.ts:121]:
```suggestion
          const state = this.getClipboardState([copyTarget], "COPY");
```
don't change the clipbaord state

File: src/plugins/ui_stateful/clipboard.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (hokolomopo) [reply src/plugins/ui_stateful/clipboard.ts]: I think your current implementation is fine. The component stays dumb and do the minimum amount of things. And the plugin re-uses the `ClipboardCellState` object so the code is really simple. My only question would be wther the command should be handeld by the Clipboard or another plugin.

But let's ask another opinion @LucasLefevre

File: src/plugins/ui_stateful/clipboard.ts (hunk 9 above)
Comment (khpa-odoo) [reply src/plugins/ui_stateful/clipboard.ts]: Applied the change.
Clipboard state won't be affected now :+1:

File: src/plugins/ui_stateful/clipboard.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (rrahir) [reply src/plugins/ui_stateful/clipboard.ts]: First things first, I am not attached to any strategy right now, I think there are theoretical pros and cons everywhere.there are some situations where the components end up wise/not-that-dumb  as they are on this lien where they want to dispatch to a plugin but there is next to 0 value added to move the logic to a plugin, except "don't put it in the component". 
I personally have some issues with the aforementioned strategy that we'll overlook right now but I feel like the commands that we design seem to me soooooo specific that we could argue that they are "action-like" (since they mimic  a user will, clearly translate a UI will).
I have come upon the same conundrum recently when wanting to implement the equivalent of 'Ctrl+Enter'  shortcut while editing a cell (it also uses the clipboard state), so facing the same issue and I am really interested by the direction we want to take :-)

File: src/plugins/ui_stateful/clipboard.ts [hunk 10]
```diff
@@ -109,6 +109,40 @@ export class ClipboardPlugin extends UIPlugin {
...
+        if (this.getters.getSelectedZones().length === 1) {
+          const zone = this.getters.getSelectedZone();
+          const singleRowSelected = zone.top !== zone.bottom;
+          const copyTarget = {
+            ...zone,
+            bottom: singleRowSelected ? zone.top : zone.top - 1,
+            top: singleRowSelected ? zone.top : zone.top - 1,
+          };
+          const state = this.getClipboardState([copyTarget], "COPY");
+          state.paste([zone], {
```
Comment (rrahir) [inline src/plugins/ui_stateful/clipboard.ts:122]: spawning a side clipboard state out of the blue feels weird

File: src/plugins/ui_stateful/clipboard.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (rrahir) [reply src/plugins/ui_stateful/clipboard.ts]: Taking a step back - we currently need a "parent" command to unify all core changes that would be dispatched by the clipboard state so the command does make sense ...

File: src/types/commands.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (rrahir) [reply src/types/commands.ts]: @hokolomopo Apologies, you are correct. `target` is reserved and should stay so.

The current sstrategy (470bec6b6005541da7d63fd44e9d3765393e6c7d) seems to have overlooked that issue by providing a custom command :)

File: src/plugins/ui_stateful/clipboard.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (khpa-odoo) [reply src/plugins/ui_stateful/clipboard.ts]: @rrahir Okay, so I moved the logic to plugin as suggested by @hokolomopo, but as far as it is concerned, I'll move it back to the component.

Also, I figure that the "parent" command would be dispatched by clipboard state. But I'm not sure how exactly it will help unifying things, since we already handle all the core changes in clipboard plugin... Can you please elaborate a bit about the way you want me to implement this?

[the remark this answers is not in the document]
Comment (rrahir) [reply src/plugins/ui_stateful/clipboard.ts]: The crucial point about the "parent" command is that we want our action (in this case the cell replication) needs to be executed in a single history step (i.e., if I undo it, every change made by the action should be reverted). There is a rule of thumb for this.  each time you call `dispatch` from outside the model (e;g. `env.model.dispatch`) it will make one history step.
On the contrary, every dispatch made from inside a plugin will be part of the very same history step,

This means that from an action, we should dispatch _once and only once_!

Now looking at the solution we have here, we want to use a `ClipboardCellsState` but this object might (and most likely will) dispatch several times. So we need this object to be used inside a plugin to have all its changes in the same history step.

This means that we need a generic command to call from outside the model, (REPLICATE_xxx) and this command will be handled inside a plugin to use the clipboard state.

tl;dr your implementation is right ATM

[the remark this answers is not in the document]
Comment (khpa-odoo) [reply src/plugins/ui_stateful/clipboard.ts]: Oh okay, I thought you probably wanted to me implement something else apart from my current implementation. 
But, thanks for clarifying and technically explaining the dispatch and history relation. I'll keep this in mind from now on :+1:

File: src/plugins/ui_stateful/clipboard.ts (hunk 10 above)
Comment (khpa-odoo) [reply src/plugins/ui_stateful/clipboard.ts:122]: @rrahir Earlier, I reused `ClipboardCellState` but then @hokolomopo pointed out that I should not be changing it's state otherwise it would mess things up in case of figure. I too, realised the same and hence I used the state as discussed [here](https://github.com/odoo/o-spreadsheet/pull/2818#discussion_r1318175545).

**Review by rrahir — CHANGES_REQUESTED** [review]
There is still a functional issue:
From a blanck spreadsheet
- add content in A1, keep the selection on A1
- Hit Ctrl +D or Ctrl+R
- A1 content dissapeared
> since we're on the edge of the spreadsheet, it should just do nothing

File: tests/clipboard/clipboard_plugin.test.ts [hunk 11]
```diff
@@ -1869,6 +1871,128 @@ describe("clipboard: pasting outside of sheet", () => {
     expect(model.getters.getSelectedZones()).toEqual([toZone("B2:AA2")]);
   });
 
+  test("Ctrl+D on cell(s) of edge row should do nothing", async () => {
```
Comment (rrahir) [inline tests/clipboard/clipboard_plugin.test.ts:1874]: don't forget to adapt the test titles ;-) the keyboard shortcuts don't belong here

Comment (khpa-odoo) [reply tests/clipboard/clipboard_plugin.test.ts]: Changed the titles :+1:

File: src/plugins/ui_stateful/clipboard.ts [hunk 12]
```diff
@@ -109,6 +109,46 @@ export class ClipboardPlugin extends UIPlugin {
           this.status = "invisible";
         }
         break;
+      case "COPY_PASTE_CELLS_ABOVE":
+        if (this.getters.getSelectedZones().length === 1) {
+          const zone = this.getters.getSelectedZone();
+          if (zone.top === 0 && zone.bottom === 0) {
```
Comment (rrahir) [inline src/plugins/ui_stateful/clipboard.ts:115]: this is typically something that we should reject in the allowDispatch rather than having a command that succeeds but does nothing

**Review by rrahir — CHANGES_REQUESTED** [review]

File: src/plugins/ui_stateful/clipboard.ts (hunk 12 above)
Comment (khpa-odoo) [reply src/plugins/ui_stateful/clipboard.ts]: Absolutely! Applied the changes :+1:

**Comment by rrahir** [comment]
@robodoo r+
